### PR TITLE
HIVE-25168: Add mutable validWriteIdList

### PR DESCRIPTION
**Pull request by hsnusonic**
* A new interface and a new class for mutable validWriteIdList
* A new test for mutable validWriteIdList

### What changes were proposed in this pull request?

Add a new mutable validWriteIdList

### Why are the changes needed?

For external metadata cache layer, it is useful to keep the validWriteIdList up-to-date by events.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

A new test added.

### CODE REVIEW

**Comment by hsnusonic** [comment]
@kishendas Could you take a look? Thanks!

File: storage-api/src/java/org/apache/hadoop/hive/common/MutableValidReaderWriteIdList.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+  public MutableValidReaderWriteIdList(ValidReaderWriteIdList writeIdList) {
+    super(writeIdList.writeToString());
+    exceptions = new ArrayList<>(exceptions);
+  }
+
+  @Override
+  public void addOpenWriteId(long writeId) {
+    if (writeId <= highWatermark) {
+      LOG.debug("Won't add any open write id because {} is less than or equal to high watermark: {}",
```
Comment (kishendas) [inline storage-api/src/java/org/apache/hadoop/hive/common/MutableValidReaderWriteIdList.java:48]: Please keep the log messages shorter, as they can occupy lot of storage space. More like ("not adding openWriteId: {} {}", writeId, highWatermark)

File: storage-api/src/java/org/apache/hadoop/hive/common/MutableValidReaderWriteIdList.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * This class is a mutable version of {@link ValidReaderWriteIdList} for use by an external cache layer.
+ * To use this class, we need to always mark the writeId as open before to mark it as aborted/committed.
+ * This class is not thread safe.
+ */
+public class MutableValidReaderWriteIdList extends ValidReaderWriteIdList implements MutableValidWriteIdList {
```
Comment (kishendas) [inline storage-api/src/java/org/apache/hadoop/hive/common/MutableValidReaderWriteIdList.java:37]: Can you add a comment on why we need this class ? More like why an external cache layer wants to use this class.

File: storage-api/src/java/org/apache/hadoop/hive/common/MutableValidReaderWriteIdList.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+import java.util.BitSet;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * This class is a mutable version of {@link ValidReaderWriteIdList} for use by an external cache layer.
+ * To use this class, we need to always mark the writeId as open before to mark it as aborted/committed.
+ * This class is not thread safe.
```
Comment (kishendas) [inline storage-api/src/java/org/apache/hadoop/hive/common/MutableValidReaderWriteIdList.java:35]: What is the implication of this class not being thread safe ?

**Pull request closed without merging**